crates/bevy_ecs/src/world/mod.rs

```diff
@@ -164,6 +164,45 @@ impl World {
         Ok(component_id)
     }
 
+    /// Registers multiple components using the given [ComponentDescriptor]s.
+    /// See [register_component](World::register_component).
+    ///
+    /// ```
+    /// use bevy_ecs::{component::{Component, ComponentDescriptor, StorageType}, world::World};
+    ///
+    /// struct Position {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// struct Velocity(f32);
+    /// struct Hostility(f32);
+    ///
+    /// // Local helper for getting a sparse set descriptor
+    /// fn cd<T: Component>() -> ComponentDescriptor {
+    ///     ComponentDescriptor::new::<T>(StorageType::SparseSet)
+    /// }
+    ///
+    /// let mut world = World::new();
+    /// let descriptors = vec![cd::<Position>(), cd::<Velocity>(), cd::<Hostility>()];
+    /// world.register_components(descriptors).unwrap();
+    /// ```
+    pub fn register_components<I>(
+        &mut self,
+        descriptors: I,
+    ) -> Result<Vec<ComponentId>, ComponentsError>
+    where
+        I: IntoIterator<Item = ComponentDescriptor>,
+    {
+        let mut ids = vec![];
```

Comment:
Can you try to pre-allocate this vec based in the iterator size_hint? Or maybe replace the function body with `descriptors.map(|descriptor| self.register_comp

Reply:
Yes, I agree with the use of `collect()` as it will allow the compiler to optimize this a little better.

Reply:
Pushed the changes from the second suggestion.

Btw. if one were to use `size_hint()`, would it have been appropriate to unwrap the upper bound and use that?
Seems the upper bound can be `None` is some cases ("infinite" iterators, or user defined iters) but those seem nonsensical for this API.

Reply:
> would it have been appropriate to unwrap the upper bound and use that?

Probably not, I could want to use my own iterator that doesn't implement `size_hint` well and that shouldn't crash